server/src/main/java/org/elasticsearch/action/support/replication/PendingReplicationActions.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support.replication;
+
+import org.elasticsearch.action.UnavailableShardsException;
+import org.elasticsearch.action.support.RetryableAction;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.index.shard.IndexShardClosedException;
+import org.elasticsearch.index.shard.ReplicationGroup;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class PendingReplicationActions implements Consumer<ReplicationGroup>, Releasable {
+
+    private final Map<String, Set<RetryableAction<?>>> onGoingReplicationActions = ConcurrentCollections.newConcurrentMap();
+    private final ShardId shardId;
+    private final ThreadPool threadPool;
+
+    public PendingReplicationActions(ShardId shardId, ThreadPool threadPool) {
+        this.shardId = shardId;
+        this.threadPool = threadPool;
+    }
+
+    public void addPendingAction(String allocationId, RetryableAction<?> replicationAction) {
+        Set<RetryableAction<?>> ongoingActionsOnNode = onGoingReplicationActions.get(allocationId);
+        if (ongoingActionsOnNode != null) {
+            ongoingActionsOnNode.add(replicationAction);
+            if (onGoingReplicationActions.containsKey(allocationId) == false) {
+                replicationAction.cancel(new UnavailableShardsException(shardId,
```

Comment:
Perhaps we can use `IndexShardClosedException` here which tells us that a shard has gone away.

Reply:
Do you want this for all the failures?